math/src/main/kotlin/ktx/math/ImmutableVector.kt

```diff
@@ -0,0 +1,154 @@
+@file:Suppress("NOTHING_TO_INLINE")
+
+package ktx.math
+
+import com.badlogic.gdx.math.Interpolation
+import com.badlogic.gdx.math.MathUtils
+import com.badlogic.gdx.math.Vector
+import kotlin.math.abs
+import kotlin.math.sqrt
+
+/**
+ * Represent an immutable vector.
+ */
+interface ImmutableVector<T : ImmutableVector<T>> : Comparable<T> {
+
+    /**
+     * Returns the squared euclidean length
+     *
+     * This method is faster than [Vector.len] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact lengths, as the return value is the square of the actual length.
+     */
+    val len2: Float
+
+    /** Returns whether the length of this vector is smaller than the given [margin] */
+    fun isZero(margin: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    /** Returns the result of subtracting the [other] vector from this vector */
+    operator fun minus(other: T): T
+
+    /** Returns the result of adding the [other] vector to this vector */
+    operator fun plus(other: T): T
+
+    /** Returns the same vector with all members incremented by 1 */
+    operator fun inc(): T
+
+    /** Returns the same vector with all members decremented by 1 */
+    operator fun dec(): T
+
+    /** Returns this vector scaled by the given [scalar] */
+    operator fun times(scalar: Float): T
+
+    /** Returns his vector scaled by the given [vector] */
+    operator fun times(vector: T): T
+
+    /** Returns he dot product of this vector by the given [vector] */
+    infix fun dot(vector: T): Float
+
+    /**
+     * Returns the squared distance between this and the other vector
+     *
+     * This method is faster than [dst] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact distance, as the return value is the square of the actual distance.
+     */
+    infix fun dst2(vector: T): Float
+
+    /** Linearly interpolates between this vector and the target vector by alpha */
+    fun lerp(target: T, alpha: Float): T
+
+    /** Returns a vector of same length and a random direction */
+    fun withRandomDirection(): T
+
+    /** Returns true if this vector is in line with the other vector (either in the same or the opposite direction) */
+    fun isOnLine(other: T, epsilon: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    /** Compares this vector with the other vector, using the supplied epsilon for fuzzy equality testing */
+    fun epsilonEquals(other: T, epsilon: Float): Boolean
+
+    override fun compareTo(other: T): Int = len2.compareTo(other.len2)
+}
+
+/**
+ * Returns the euclidean length
+ */
+inline val ImmutableVector<*>.len: Float get() = sqrt(len2)
+
+/**
+ * Returns the unit vector of same direction or this vector if it is zero.
+ */
+val <T : ImmutableVector<T>> T.nor: T
+    get() {
+        val l2 = len2
+
+        if (l2 == 1f || l2 == 0f) return this
```

Comment:
Convert to `if`+`else` expression, return it right away. This isn't very idiomatic. Same goes for the method below.

Why is this property outside of the interface?

Reply:
>Convert to if+else expression, return it right away. This isn't very idiomatic. Same goes for the method below.

Ok

>Why is this property outside of the interface?

Because it is only an helper available for any `ImmutableVector`. Writing it outside of the interface bring to advantages:
* No need to rewrite it for Vector3
* Future proof (It is easy and safe to move from extension to member, not the opposite)

Reply:
I see a single reason why you would want these methods to be within the interface - you can provide your own implementation. Most immutable vector methods are deterministic and depend on the immutable state of the vector (e.g. length, angle), so in some cases people might want to cache these values e.g. by introducing a new lazy field. I prefer thin interfaces as well, so I won't try to argue, but consider if any of the properties/methods you introduced outside of the interface or class definitions might need overriding.

Reply:
>you can provide your own implementation

This would indeed be a very good reason to make this/theses function member of the interface. I wasn't able to think of a specific implementation and making it a member function in the future will be possible without breaking API. (members always have precedence over extension functions).

But I tell all that only to share my thinking. I don't think there's anything to argue. I will, of course, align to your preferences. Shall I make it a member then? If yes, all the functions or only the properties?

Reply:
There are other ways of caching results from vectors' methods and I think `ImmutableVector2` currently cannot be extended either way. I don't see how most of these operations could need a different implementation that returns a different value either (e.g. length).

It's a tradeoff between keeping the API close to LibGDX (huge interface), and clean/idiomatic. I'll accept the PR with either approach, just go through all of the methods and consider if any of them might need to be more generic or extendable:

> (...) consider if any of the properties/methods you introduced outside of the interface or class definitions might need overriding.

Reply:
I agree with everything you said. I'll go through all of them and promote some of them to be member (like `len`).

As a side note I observed that a lot of Java libraries choose to provide "Huge interfaces" as a trade-off to not have to use helper classes, which is unpleasant for the user (this is the case of `Observable` in Rx for instance). But we don't have this limitation in Kotlin :-p)